content/blog/2019-10-04-InstructionScheduling.md

```diff
@@ -0,0 +1,143 @@
+
+
+### Computational Complexity of the Algorithm
+Let the number of instructions in the basic block be n.
+#### Constructing the dag
+In the worst case, every instruction needs to be compared with the instructions already in dag. During this process, we can also compute the information needed by the heuristic evaluation, such as the length of the longest path and the number of immediate successors. It is thus O(n<sup>2</sup>).
+#### Scheduling the instructions
+To schedule an instruction, we need to evaluate all the candidates based on the heuristic. Evaluation can be done in O(1) because we already build the dag contains all the relevant information. Thus, scheduling is also O(n<sup>2</sup>).
+
+#### Overall complexity
+The overall complexity is O(n<sup>2</sup>). This is significantly better compared to other algorithms in this space which are O(n<sup>4</sup>), even when adjusted to have similar hardware assumptions.
+
+### Experiments
+The authors implemented this instruction scheduler and made the following observations from benchmark results:
+ - In practice, these heuristics effectively remove avoidable interlocks and runs in approximately linear time.
+ - The memory referencing assumptions greatly improve results, and effectiveness increases with better aliasing information (provided by other parts of the compiler).
+ - Carry/Borrow subgraph do not improve much for most programs. Significant improvements only occur when the program is computationally intensive. 
+ - Using more versatile dags proposed by other literature only slightly improves the instruction scheduling effectiveness. 
+
+The referenced additional information on performance in the [[Joh86]](https://dl.acm.org/citation.cfm?id=13321) paper which tested load/store scheduling and showed a 5% improvement. This improvement was measured by the reduction in interlocks caused by load/store instructions. 
+
+## Our Thoughts
+This paper left a few things to be desired. 
+
+First, we did not quite understand some of the hazards this paper was concerned about. The first hazard of "loading a register from memory followed by using *that* register as a source" is clear as that's a traditional load-use hazard where the load finishes at the end of the cycle while the value it was loading was needed at the beginning of the cycle. 
+
+However, the second hazard of "storing to any memory location followed by loading from any location" was puzzling as it is not clear why this would not work. Even if we considered the addresses to alias and assumed no store-to-load forwarding, it seems reasonable that the store would finish before the subsequent load and that the load would return the correct value. If we go a step further and assume memory accesses take multiple cycles, there still does not appear to be a problem as this is a read-after-write dependency, and the writing happens before the read. 
```

Comment:
That one is a bit trickier, but I bet their architecture can do memory accesses at different points in the pipeline. For example, it seems possible to do a memory op as part of an "add" instruction, which means the memory read takes place before the execute stage. That means that a subsequent write to memory would happen later in the pipeline, causing a stall.